### PR TITLE
Special Treatment of Regions in DaCe backends

### PR DESCRIPTION
**Pull request by gronerl**
This PR implements a expansion configuration where
* for HorizontalExecutions with a single region, the regions are implemented by restriting the map ranges instead of guarding statements with a predicate.
* HorizontalExecutions where each statement is only applied to a corner of a domain are broken up into multiple HorizontalExecutions (making them amenable to the much more efficient implementation of the first bullet point.)

### CODE REVIEW

**Comment by gronerl** [comment]
bors try

File: src/gtc/dace/expansion/daceir_builder.py [hunk 1]
```diff
@@ -103,6 +105,38 @@ def _get_tasklet_inout_memlets(node: oir.HorizontalExecution, *, get_outputs, gl
...
 
+def _all_stmts_same_region(scope_nodes, axis: dcir.Axis, interval):
+    return (
+        axis in dcir.Axis.dims_horizontal()
+        and isinstance(interval, dcir.DomainInterval)
+        and all(
+            isinstance(stmt, oir.MaskStmt) and isinstance(stmt.mask, common.HorizontalMask)
+            for tasklet in iter_tree(scope_nodes).if_isinstance(dcir.Tasklet)
+            for stmt in tasklet.stmts
+        )
```
Comment (jdahm) [inline src/gtc/dace/expansion/daceir_builder.py:116]: I like this, but for readability in the future, could you make a variable for each of these conditions with a name that describes it then `return x and y and z`?

File: src/gtc/dace/expansion/utils.py [hunk 2]
```diff
@@ -32,3 +43,127 @@ def get_dace_debuginfo(node: common.LocNode):
...
+        mask_attrs[axis.lower()] = self.visit(getattr(node, axis.lower()))
+        return common.HorizontalMask(**mask_attrs)
+
+    def visit_HorizontalInterval(self, node: common.HorizontalInterval):
+        return common.HorizontalInterval(start=None, end=None)
+
+
+class HorizontalMaskRemover(eve.NodeMutator):
+    def visit_Tasklet(self, node: "dcir.Tasklet"):
+        from gtc import daceir as dcir
```
[the remark this answers is not in the document]
Comment (jdahm) [reply src/gtc/dace/expansion/utils.py:60]: Ah I see `import daceir as daceir` at the top when `TYPE_CHECKING`. Is there a circular import otherwise?

**Review by jdahm — APPROVED** [review]
Looks good. I just gave a few comments. I can't tell if there's actually a test that the expansion works for the region, but I guess if it produces the correct result on the test suite that's sufficient.

**Comment by gronerl** [comment]
bors try

**Comment by gronerl** [comment]
bors try

File: src/gtc/dace/expansion/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gronerl) [reply src/gtc/dace/expansion/utils.py]: Likely was at some point. Not anymore it seems :thinking:

File: src/gtc/dace/expansion/daceir_builder.py (hunk 1 above)
Comment (gronerl) [reply src/gtc/dace/expansion/daceir_builder.py]: done

**Pull request merged**